File: pysm/semantic_modeling/assembling/weak_models/local_structure.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
from typing import Dict, Tuple, List

from semantic_modeling.assembling.weak_models.multi_val_predicate import MultiValuePredicate
from semantic_modeling.assembling.weak_models.statistic import Statistic
from semantic_modeling.data_io import get_semantic_models
from semantic_modeling.karma.semantic_model import SemanticModel


class NodeStructureSpace:

    def __init__(self, lbl: bytes,
                 parents: Dict[Tuple[bytes, bytes], int],
                 children: Dict[Tuple[bytes, bytes], int],
                 is_class_child: List[bool],
                 child_usage_count: List[int]) -> None:
        self.lbl = lbl
        self.parents = parents
        self.children = children

        # a map from child idx => is_class_node
        self.is_class_child = is_class_child

        # a map from child_idx => count
        self.child_usage_count = child_usage_count

    def get_parent_idx(self, link_lbl: bytes, parent_lbl: bytes):
        return self.parents.get((link_lbl, parent_lbl), None)

    def get_child_idx(self, link_lbl: bytes, child_lbl: bytes):
        return self.children.get((link_lbl, child_lbl), None)


class LocalStructure(object):

    instance = None

    def __init__(self, train_sms: List[SemanticModel]) -> None:
        self.train_sm_ids = {sm.id for sm in train_sms}
        self.node_structure_space: Dict[bytes, NodeStructureSpace] = {}

        # MAKE RAW STRUCTURE SPACE FIRST
        raw_node_structure_space: Dict[bytes, dict] = {}
        for sm in train_sms:
            for n in sm.graph.iter_class_nodes():
                if n.label not in raw_node_structure_space:
                    raw_node_structure_space[n.label] = {
                        "parents": {},
                        "children": {},
                    }

                plink = n.get_first_incoming_link()
                if plink is not None:
                    triple = (plink.label, plink.get_source_node().label)
                    if triple not in raw_node_structure_space[n.label]['parents']:
                        raw_node_structure_space[n.label]['parents'][triple] = 0
                    raw_node_structure_space[n.label]['parents'][triple] += 1

                for e in n.iter_outgoing_links():
                    target = e.get_target_node()
                    triple = (e.label, b"DATA_NODE" if target.is_data_node() else target.label)
                    assert target.is_data_node()  or target.label != b"DATA_NODE"
                    if triple not in raw_node_structure_space[n.label]['children']:
                        raw_node_structure_space[n.label]['children'][triple] = 0
                    raw_node_structure_space[n.label]['children'][triple] += 1

        # MAKE NODE STRUCTURE SPACE
        for n, space in raw_node_structure_space.items():
            children_attrs = sorted(space['children'].keys(), key=lambda x: space['children'][x], reverse=True)
            self.node_structure_space[n] = NodeStructureSpace(
                n, {x: i for i, x in enumerate(space['parents'].keys())},
                {x: i for i, x in enumerate(children_attrs)},
                [x[1] != b'DATA_NODE' for x in children_attrs],
                [space['children'][x] for x in children_attrs])

    @staticmethod
    def get_instance(train_sms: List[SemanticModel]) -> 'LocalStructure':
        sm_ids = {sm.id for sm in train_sms}

        if LocalStructure.instance is None:
            LocalStructure.instance = LocalStructure(train_sms)
            return LocalStructure.instance

        assert LocalStructure.instance.train_sm_ids == sm_ids
        return LocalStructure.instance


if __name__ == '__main__':
    import ujson

    dataset = "museum_edm"
    train_size = 14
    source_models = get_semantic_models(dataset)[:train_size]

    local_structure = LocalStructure.get_instance(source_models)
    print(ujson.dumps(local_structure.node_structure_space, indent=4))

File: mira/src/assembling/features/local_structure.rs
use assembling::features::primary_key::PrimaryKey;
use itertools::Itertools;
use std::collections::HashMap;
use rdb2rdf::models::semantic_model::SemanticModel;
use std::ops::AddAssign;
use std::cmp::Ordering;

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct NodeStructureSpace {
    // parent_idx and child_idx always different, parent_idx < child_idx
    pub label: String,
    pub parents: HashMap<(String, String), usize>,
    pub children: HashMap<(String, String), usize>,
    pub child_pred2idx: HashMap<String, usize>,
    is_class_child: Vec<bool>,
    child_usage_count: Vec<usize>
}

impl NodeStructureSpace {
    pub fn get_parent_idx(&self, link_lbl: &str, parent_lbl: &str) -> Option<&usize> {
        self.parents.get(&(link_lbl.to_owned(), parent_lbl.to_owned()))
    }

    pub fn get_child_idx(&self, link_lbl: &str, child_lbl: &str) -> Option<&usize> {
        self.children.get(&(link_lbl.to_owned(), child_lbl.to_owned()))
    }

    pub fn has_smaller_index(&self, link_lbl_a: &str, link_lbl_b: &str) -> bool {
        self.child_pred2idx[link_lbl_a] < self.child_pred2idx[link_lbl_b]
    }
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct LocalStructure {
    pub node_structure_space: HashMap<String, NodeStructureSpace>
}

#[derive(Debug)]
struct RawNodeStruct<'a> {
    parents: HashMap<(&'a str, &'a str), usize>,
    children: HashMap<(&'a str, &'a str), usize>,
}

impl Default for LocalStructure {
    fn default() -> Self {
        LocalStructure {
            node_structure_space: Default::default()
        }
    }
}

impl LocalStructure {
    pub fn new(train_sms: &[&SemanticModel], primary_key: &PrimaryKey) -> LocalStructure {
        let mut node_structure_space: HashMap<String, NodeStructureSpace> = Default::default();

        // make raw structure space first
        let mut raw_node_structure_space: HashMap<&str, RawNodeStruct> = Default::default();
        for sm in train_sms {
            for n in sm.graph.iter_class_nodes() {
                if !raw_node_structure_space.contains_key(n.label.as_str()) {
                    raw_node_structure_space.insert(&n.label, RawNodeStruct {
                        parents: Default::default(),
                        children: Default::default(),
                    });
                }

                let p_edge = n.first_incoming_edge(&sm.graph);
                if let Some(p_edge) = p_edge {
                    let triple = (p_edge.label.as_str(), p_edge.get_source_node(&sm.graph).label.as_str());
                    raw_node_structure_space.get_mut(n.label.as_str()).unwrap()
                        .parents.entry(triple).or_insert(0).add_assign(1);
                }

                for e in n.iter_outgoing_edges(&sm.graph).unique_by(|e| &e.label) {
                    let target = e.get_target_node(&sm.graph);
                    let triple = (e.label.as_str(), if target.is_data_node() {
                        "DATA_NODE"
                    } else {
                        &target.label
                    });

                    raw_node_structure_space.get_mut(n.label.as_str()).unwrap()
                        .children.entry(triple).or_insert(0).add_assign(1);
                }
            }
        }

        // make node structure space
        for (&n_lbl, space) in raw_node_structure_space.iter_mut() {
            let pk = if primary_key.contains(n_lbl) {
                primary_key.get_primary_key(n_lbl)
            } else {
                "__no_pk__"
            };

            let mut children_attrs = space.children.keys().collect::<Vec<_>>();
            children_attrs.sort_by(|a, b| {
                // always make primary key the first element!!
                if a.0 == pk {
                    Ordering::Less
                } else if b.0 == pk {
                    Ordering::Greater
                } else if space.children[b] == space.children[a] && a.0 != b.0 {
                    // compare by lexical order
                    b.0.cmp(&a.0)
                } else {
                    space.children[b].cmp(&space.children[a])
                }
            });

            let child_pred2idx = children_attrs.iter()
                .unique_by(|a| a.0)
                .enumerate()
                .map(|(i, a)| (a.0.to_owned(), i))
                .collect::<HashMap<_, _>>();

            node_structure_space.insert(n_lbl.to_owned(), NodeStructureSpace {
                label: n_lbl.to_owned(),
                parents: space.parents.keys().enumerate()
                    .map(|(i, v)| ((v.0.to_owned(), v.1.to_owned()), i))
                    .collect(),
                children: children_attrs.iter().enumerate()
                    .map(|(i, v)| ((v.0.to_owned(), v.1.to_owned()), i + space.parents.len()))
                    .collect(),
                child_pred2idx,
                is_class_child: children_attrs.iter().map(|v| v.1 != "DATA_NODE").collect(),
                child_usage_count: children_attrs.iter().map(|v| space.children[v]).collect()
            });
        }

        LocalStructure {
            node_structure_space
        }
    }
}